File: backend/app/schemas/tesoreria.py
"""
Schemas de Tesorería.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field


# ==================== CHEQUE ====================

class ChequeCreate(BaseModel):
    """Schema para crear un cheque."""
    numero: str = Field(..., min_length=1, max_length=50)
    tipo: str = Field(default="fisico")  # fisico, echeq
    origen: str = Field(default="recibido_cliente")  # recibido_cliente, recibido_proveedor, emitido

    monto: Decimal = Field(..., gt=0)
    fecha_emision: Optional[date] = None
    fecha_vencimiento: date

    banco_origen: Optional[str] = Field(None, max_length=100)
    cuenta_destino_id: Optional[UUID] = None
    banco_destino: Optional[str] = Field(None, max_length=100)

    cliente_id: Optional[UUID] = None
    proveedor_id: Optional[UUID] = None

    librador: Optional[str] = Field(None, max_length=200)
    cuit_librador: Optional[str] = Field(None, max_length=15)

    notas: Optional[str] = None


class ChequeUpdate(BaseModel):
    """Schema para actualizar un cheque."""
    numero: Optional[str] = Field(None, min_length=1, max_length=50)
    tipo: Optional[str] = None
    estado: Optional[str] = None

    monto: Optional[Decimal] = Field(None, gt=0)
    fecha_emision: Optional[date] = None
    fecha_vencimiento: Optional[date] = None
    fecha_cobro: Optional[date] = None

    banco_origen: Optional[str] = Field(None, max_length=100)
    cuenta_destino_id: Optional[UUID] = None
    banco_destino: Optional[str] = Field(None, max_length=100)

    cliente_id: Optional[UUID] = None
    proveedor_id: Optional[UUID] = None

    librador: Optional[str] = Field(None, max_length=200)
    cuit_librador: Optional[str] = Field(None, max_length=15)

    notas: Optional[str] = None
    motivo_rechazo: Optional[str] = None


class ChequeResponse(BaseModel):
    """Schema de respuesta de cheque."""
    id: UUID
    numero: str
    tipo: str
    origen: str
    estado: str

    monto: Decimal
    fecha_emision: Optional[date]
    fecha_vencimiento: date
    fecha_cobro: Optional[date]

    banco_origen: Optional[str]
    cuenta_destino_id: Optional[UUID]
    banco_destino: Optional[str]

    cliente_id: Optional[UUID]
    proveedor_id: Optional[UUID]

    librador: Optional[str]
    cuit_librador: Optional[str]

    registrado_por_id: UUID
    cobrado_por_id: Optional[UUID]

    notas: Optional[str]
    motivo_rechazo: Optional[str]

    created_at: datetime
    activo: bool

    # Campos calculados
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None
    registrado_por_nombre: Optional[str] = None
    dias_para_vencimiento: Optional[int] = None

    class Config:
        from_attributes = True


class ChequeList(BaseModel):
    """Schema para listado de cheques."""
    id: UUID
    numero: str
    tipo: str
    origen: str
    estado: str
    monto: Decimal
    fecha_vencimiento: date
    fecha_cobro: Optional[date]
    banco_origen: Optional[str]
    banco_destino: Optional[str]
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None
    dias_para_vencimiento: Optional[int] = None

    class Config:
        from_attributes = True


class DepositarChequeRequest(BaseModel):
    """Schema para depositar un cheque."""
    cuenta_destino_id: UUID
    fecha_deposito: date = Field(default_factory=date.today)
    notas: Optional[str] = None


class CobrarChequeRequest(BaseModel):
    """Schema para marcar cheque como cobrado."""
    fecha_cobro: date = Field(default_factory=date.today)
    notas: Optional[str] = None


class RechazarChequeRequest(BaseModel):
    """Schema para rechazar un cheque."""
    motivo_rechazo: str = Field(..., min_length=3)
    fecha_rechazo: date = Field(default_factory=date.today)


class EntregarChequeRequest(BaseModel):
    """Schema para entregar cheque a tercero."""
    proveedor_id: Optional[UUID] = None
    concepto: str = Field(..., min_length=3)
    fecha_entrega: date = Field(default_factory=date.today)
    notas: Optional[str] = None


# ==================== MOVIMIENTO TESORERIA ====================

class MovimientoTesoreriaCreate(BaseModel):
    """Schema para crear movimiento de tesorería."""
    tipo: str  # ingreso_efectivo, ingreso_transferencia, etc.
    concepto: str = Field(..., min_length=3, max_length=200)
    descripcion: Optional[str] = None

    monto: Decimal = Field(..., gt=0)
    es_ingreso: bool = True
    fecha_movimiento: date
    fecha_valor: Optional[date] = None

    metodo_pago: str  # efectivo, transferencia, cheque

    # Si es transferencia
    banco_origen: Optional[str] = None
    banco_destino: Optional[str] = None
    cuenta_destino_id: Optional[UUID] = None
    numero_transferencia: Optional[str] = None

    # Si es cheque existente
    cheque_id: Optional[UUID] = None

    # Relaciones
    cliente_id: Optional[UUID] = None
    proveedor_id: Optional[UUID] = None

    notas: Optional[str] = None
    comprobante: Optional[str] = None


class MovimientoTesoreriaResponse(BaseModel):
    """Schema de respuesta de movimiento de tesorería."""
    id: UUID
    tipo: str
    concepto: str
    descripcion: Optional[str]

    monto: Decimal
    es_ingreso: bool
    fecha_movimiento: date
    fecha_valor: Optional[date]

    metodo_pago: str
    banco_origen: Optional[str]
    banco_destino: Optional[str]
    cuenta_destino_id: Optional[UUID]
    numero_transferencia: Optional[str]

    cheque_id: Optional[UUID]
    cliente_id: Optional[UUID]
    proveedor_id: Optional[UUID]

    registrado_por_id: UUID
    notas: Optional[str]
    comprobante: Optional[str]

    anulado: bool
    motivo_anulacion: Optional[str]

    created_at: datetime
    activo: bool

    # Campos calculados
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None
    registrado_por_nombre: Optional[str] = None
    cheque_numero: Optional[str] = None

    class Config:
        from_attributes = True


class MovimientoTesoreriaList(BaseModel):
    """Schema para listado de movimientos."""
    id: UUID
    tipo: str
    concepto: str
    monto: Decimal
    es_ingreso: bool
    fecha_movimiento: date
    metodo_pago: str
    cliente_nombre: Optional[str] = None
    proveedor_nombre: Optional[str] = None
    cheque_numero: Optional[str] = None
    anulado: bool

    class Config:
        from_attributes = True


class AnularMovimientoRequest(BaseModel):
    """Schema para anular movimiento."""
    motivo: str = Field(..., min_length=5)


# ==================== RESUMEN ====================

class ResumenTesoreria(BaseModel):
    """Resumen de tesorería."""
    # Cheques
    cheques_en_cartera: int
    total_cheques_cartera: Decimal
    cheques_proximos_vencer: int  # próximos 7 días
    total_proximos_vencer: Decimal
    cheques_vencidos: int
    total_vencidos: Decimal

    # Movimientos del período
    total_ingresos_efectivo: Decimal
    total_ingresos_transferencia: Decimal
    total_ingresos_cheque: Decimal
    total_egresos_efectivo: Decimal
    total_egresos_transferencia: Decimal
    total_egresos_cheque: Decimal

    # Saldo
    saldo_periodo: Decimal


# ==================== CONSTANTES ====================

from app.models.tesoreria import (
    TIPOS_CHEQUE,
    ORIGENES_CHEQUE,
    ESTADOS_CHEQUE,
    METODOS_PAGO_TESORERIA,
    BANCOS_ARGENTINA,
)

File: frontend/src/types/tesoreria.ts
/**
 * Tipos para el módulo de Tesorería
 */

// ==================== CHEQUE ====================

export type TipoCheque = 'fisico' | 'echeq';
export type OrigenCheque = 'recibido_cliente' | 'recibido_proveedor' | 'emitido';
export type EstadoCheque = 'en_cartera' | 'depositado' | 'cobrado' | 'entregado' | 'rechazado' | 'anulado';

export interface Cheque {
  id: string;
  numero: string;
  tipo: TipoCheque;
  origen: OrigenCheque;
  estado: EstadoCheque;
  monto: number;
  fecha_emision: string | null;
  fecha_vencimiento: string;
  fecha_cobro: string | null;
  banco_origen: string | null;
  cuenta_destino_id: string | null;
  banco_destino: string | null;
  cliente_id: string | null;
  proveedor_id: string | null;
  librador: string | null;
  cuit_librador: string | null;
  registrado_por_id: string;
  cobrado_por_id: string | null;
  notas: string | null;
  motivo_rechazo: string | null;
  created_at: string;
  activo: boolean;
  // Campos calculados
  cliente_nombre: string | null;
  proveedor_nombre: string | null;
  registrado_por_nombre: string | null;
  dias_para_vencimiento: number | null;
}

export interface ChequeList {
  id: string;
  numero: string;
  tipo: TipoCheque;
  origen: OrigenCheque;
  estado: EstadoCheque;
  monto: number;
  fecha_vencimiento: string;
  fecha_cobro: string | null;
  banco_origen: string | null;
  banco_destino: string | null;
  cliente_nombre: string | null;
  proveedor_nombre: string | null;
  dias_para_vencimiento: number | null;
}

export interface ChequeCreate {
  numero: string;
  tipo?: TipoCheque;
  origen?: OrigenCheque;
  monto: number;
  fecha_emision?: string | null;
  fecha_vencimiento: string;
  banco_origen?: string | null;
  cuenta_destino_id?: string | null;
  banco_destino?: string | null;
  cliente_id?: string | null;
  proveedor_id?: string | null;
  librador?: string | null;
  cuit_librador?: string | null;
  notas?: string | null;
}

export interface ChequeUpdate {
  numero?: string;
  tipo?: TipoCheque;
  estado?: EstadoCheque;
  monto?: number;
  fecha_emision?: string | null;
  fecha_vencimiento?: string;
  fecha_cobro?: string | null;
  banco_origen?: string | null;
  cuenta_destino_id?: string | null;
  banco_destino?: string | null;
  cliente_id?: string | null;
  proveedor_id?: string | null;
  librador?: string | null;
  cuit_librador?: string | null;
  notas?: string | null;
  motivo_rechazo?: string | null;
}

export interface DepositarChequeRequest {
  cuenta_destino_id: string;
  fecha_deposito?: string;
  notas?: string | null;
}

export interface CobrarChequeRequest {
  fecha_cobro?: string;
  notas?: string | null;
}

export interface RechazarChequeRequest {
  motivo_rechazo: string;
  fecha_rechazo?: string;
}

export interface EntregarChequeRequest {
  proveedor_id?: string | null;
  concepto: string;
  fecha_entrega?: string;
  notas?: string | null;
}

// ==================== MOVIMIENTO TESORERIA ====================

export type TipoMovimientoTesoreria =
  | 'ingreso_efectivo'
  | 'ingreso_transferencia'
  | 'ingreso_cheque'
  | 'egreso_efectivo'
  | 'egreso_transferencia'
  | 'egreso_cheque'
  | 'deposito_cheque'
  | 'cobro_cheque';

export type MetodoPagoTesoreria = 'efectivo' | 'transferencia' | 'cheque';

export interface MovimientoTesoreria {
  id: string;
  tipo: string;
  concepto: string;
  descripcion: string | null;
  monto: number;
  es_ingreso: boolean;
  fecha_movimiento: string;
  fecha_valor: string | null;
  metodo_pago: MetodoPagoTesoreria;
  banco_origen: string | null;
  banco_destino: string | null;
  cuenta_destino_id: string | null;
  numero_transferencia: string | null;
  cheque_id: string | null;
  cliente_id: string | null;
  proveedor_id: string | null;
  registrado_por_id: string;
  notas: string | null;
  comprobante: string | null;
  anulado: boolean;
  motivo_anulacion: string | null;
  created_at: string;
  activo: boolean;
  // Campos calculados
  cliente_nombre: string | null;
  proveedor_nombre: string | null;
  registrado_por_nombre: string | null;
  cheque_numero: string | null;
}

export interface MovimientoTesoreriaList {
  id: string;
  tipo: string;
  concepto: string;
  monto: number;
  es_ingreso: boolean;
  fecha_movimiento: string;
  metodo_pago: MetodoPagoTesoreria;
  cliente_nombre: string | null;
  proveedor_nombre: string | null;
  cheque_numero: string | null;
  anulado: boolean;
}

export interface MovimientoTesoreriaCreate {
  tipo: string;
  concepto: string;
  descripcion?: string | null;
  monto: number;
  es_ingreso: boolean;
  fecha_movimiento: string;
  fecha_valor?: string | null;
  metodo_pago: MetodoPagoTesoreria;
  banco_origen?: string | null;
  banco_destino?: string | null;
  cuenta_destino_id?: string | null;
  numero_transferencia?: string | null;
  cheque_id?: string | null;
  cliente_id?: string | null;
  proveedor_id?: string | null;
  notas?: string | null;
  comprobante?: string | null;
}

export interface AnularMovimientoRequest {
  motivo: string;
}

// ==================== MOVIMIENTO CONSOLIDADO ====================

export interface MovimientoConsolidado {
  id: string;
  fecha: string;
  tipo: string;
  origen: 'cheque' | 'movimiento_tesoreria' | 'movimiento_bancario';
  concepto: string;
  monto: number;
  es_ingreso: boolean;
  metodo_pago: string | null;
  cliente_id: string | null;
  cliente_nombre: string | null;
  proveedor_id: string | null;
  proveedor_nombre: string | null;
  numero_referencia: string | null;
  banco: string | null;
  estado: string | null;
  created_at: string;
}

export interface MovimientosConsolidadosResponse {
  items: MovimientoConsolidado[];
  total: number;
  total_ingresos: number;
  total_egresos: number;
  skip: number;
  limit: number;
}

// ==================== RESUMEN ====================

export interface ResumenTesoreria {
  cheques_en_cartera: number;
  total_cheques_cartera: number;
  cheques_proximos_vencer: number;
  total_proximos_vencer: number;
  cheques_vencidos: number;
  total_vencidos: number;
  total_ingresos_efectivo: number;
  total_ingresos_transferencia: number;
  total_ingresos_cheque: number;
  total_egresos_efectivo: number;
  total_egresos_transferencia: number;
  total_egresos_cheque: number;
  saldo_periodo: number;
}

// ==================== CONSTANTES ====================

export const TIPOS_CHEQUE = [
  { value: 'fisico', label: 'Cheque Físico' },
  { value: 'echeq', label: 'E-Cheq' },
];

export const ORIGENES_CHEQUE = [
  { value: 'recibido_cliente', label: 'Recibido de Cliente' },
  { value: 'recibido_proveedor', label: 'Recibido de Proveedor' },
  { value: 'emitido', label: 'Emitido (Propio)' },
];

export const ESTADOS_CHEQUE = [
  { value: 'en_cartera', label: 'En Cartera', color: 'blue' },
  { value: 'depositado', label: 'Depositado', color: 'yellow' },
  { value: 'cobrado', label: 'Cobrado', color: 'green' },
  { value: 'entregado', label: 'Entregado', color: 'purple' },
  { value: 'rechazado', label: 'Rechazado', color: 'red' },
  { value: 'anulado', label: 'Anulado', color: 'gray' },
];

export const METODOS_PAGO_TESORERIA = [
  { value: 'efectivo', label: 'Efectivo' },
  { value: 'transferencia', label: 'Transferencia' },
  { value: 'cheque', label: 'Cheque' },
];

export const BANCOS_ARGENTINA = [
  'Banco Nación',
  'Banco Provincia',
  'Banco Ciudad',
  'Banco Galicia',
  'Banco Santander',
  'BBVA',
  'Banco Macro',
  'Banco Patagonia',
  'HSBC',
  'Banco ICBC',
  'Banco Credicoop',
  'Banco Hipotecario',
  'Banco Supervielle',
  'Banco Comafi',
  'Banco de Córdoba',
  'Banco de San Juan',
  'Brubank',
  'Mercado Pago',
  'Naranja X',
  'Ualá',
  'Otro',
];
